### PR TITLE
Fixed error reported from bind for valid IN6_ADDR when setting TTL/HOPS or TOS/TCLASS

### LINKED ISSUE #810: srt_bind throws exception for valid IN6_ADDR when setting TTL/HOPS or TOS/TCLASS
Relates to commit ee311d6ac10a008a5a27f079cbc22463f8565690

srt_bind() ends up calling CChannel::setUDPSockOpt().  If the socket is an AF_INET6 socket and the user has specified SRTO_IPTTL or SRTO_IPTOS options, this function will try to set these options in both IPPROTO_IP (IPv4) and IPPROTO_IP6 (IPv6), assuming a dual protocol stack, so that the listening socket can receive messages from an IPv4 or IPv6 remote sender.  However the internal call to ::setsockopt() returns an error for some important scenarios, resulting in an exception thrown.

Reported on Windows 10 Enterprise version 1703 

The current code does work for setting SRTO_IPTTL if the IN6_ADDR associated with the socket is unspecified.  This is the case if getaddrinfo() is not called to set the bind address, or if getaddrinfo() is called with an empty string for the network interface address.  In these cases, the IN6_ADDR is [::], and the srt_bind() operation results in listening on both 0.0.0.0 (all IPv4 interfaces) and [::] (all IPv6 interfaces).  

The problem happens when you try to bind the socket to a specific interface.  For example, if the IN6_ADDR associated with the socket is a pure IPv6 socket such as [::1] (localhost) then the call to ::setsockopt() with IPPROTO_IP will fail with WIndows Sockets Error Code WSAEINVAL 10022 Invalid argument.

Likewise if the IN6_ADDR is an IPv4-mapped IPv6 address such as [::FFFF:127.0.0.1] (localhost) then the call to ::setsockopt() with IPPROTO_IP6 will fail with the same error code.

I think the current code does not work for setting SRTO_IPTOS in any case.  Check the #ifdef IPV6_TCLASS around line 230.  The matching #endif appears to be in the wrong place.

Here is the code we're using to fix these issues, replacing the code around lines 214 and 230 in channel.cpp.  It works in all the cases we have tested.

```
#ifdef SRT_ENABLE_IPOPTS
      if (-1 != m_iIpTTL)
      {
         if(m_iIPversion == AF_INET)
         {
            if(0 != ::setsockopt(m_iSocket, IPPROTO_IP, IP_TTL, (const char*)&m_iIpTTL, sizeof(m_iIpTTL)))
               throw CUDTException(MJ_SETUP, MN_NORES, NET_ERROR);
         }
         else //Assuming AF_INET6
         {
            if (m_BindAddr.family() == AF_INET6) {
               if (IN6_IS_ADDR_UNSPECIFIED(&m_BindAddr.sin6.sin6_addr) || !IN6_IS_ADDR_V4MAPPED(&m_BindAddr.sin6.sin6_addr)) {
                  if (0 != ::setsockopt(m_iSocket, IPPROTO_IPV6, IPV6_UNICAST_HOPS, (const char*)&m_iIpTTL, sizeof(m_iIpTTL))) {
                     throw CUDTException(MJ_SETUP, MN_NORES, NET_ERROR);
                  }
               }
               if (IN6_IS_ADDR_UNSPECIFIED(&m_BindAddr.sin6.sin6_addr) || IN6_IS_ADDR_V4MAPPED(&m_BindAddr.sin6.sin6_addr)) {
                  //For IPv4mapped-IPv6 accepted connection also set the IPV4 socket.
                  if (0 != ::setsockopt(m_iSocket, IPPROTO_IP, IP_TTL, (const char*)&m_iIpTTL, sizeof(m_iIpTTL))) {
                     throw CUDTException(MJ_SETUP, MN_NORES, NET_ERROR);
                  }
               }
            }
         }
      }
      if (-1 != m_iIpToS)
      {
         if(m_iIPversion == AF_INET)
         {
            if(0 != ::setsockopt(m_iSocket, IPPROTO_IP, IP_TOS, (const char*)&m_iIpToS, sizeof(m_iIpToS)))
               throw CUDTException(MJ_SETUP, MN_NORES, NET_ERROR);
         }
         else //Assuming AF_INET6
         {
            if (m_BindAddr.family() == AF_INET6) {
               if (IN6_IS_ADDR_UNSPECIFIED(&m_BindAddr.sin6.sin6_addr) || !IN6_IS_ADDR_V4MAPPED(&m_BindAddr.sin6.sin6_addr)) {
#ifdef IPV6_TCLASS
                  if (0 != ::setsockopt(m_iSocket, IPPROTO_IPV6, IPV6_TCLASS, (const char*)&m_iIpToS, sizeof(m_iIpToS))) {
                     throw CUDTException(MJ_SETUP, MN_NORES, NET_ERROR);
                  }
#endif
               }
               if (IN6_IS_ADDR_UNSPECIFIED(&m_BindAddr.sin6.sin6_addr) || IN6_IS_ADDR_V4MAPPED(&m_BindAddr.sin6.sin6_addr)) {
                  //For IPv4mapped-IPv6 accepted connection also set the IPV4 socket.
                  if (0 != ::setsockopt(m_iSocket, IPPROTO_IP, IP_TOS, (const char*)&m_iIpToS, sizeof(m_iIpToS))) {
                     throw CUDTException(MJ_SETUP, MN_NORES, NET_ERROR);
                  }
               }
            }
         }
      }
#endif
```

### PR DESCRIPTION
**Pull request by ethouris**
Fixes #810

### CODE REVIEW

File: srtcore/channel.cpp [hunk 1]
```diff
@@ -201,40 +202,70 @@ void CChannel::setUDPSockOpt()
...
-         else //Assuming AF_INET6
-         {
-            if(0 != ::setsockopt(m_iSocket, IPPROTO_IPV6, IPV6_UNICAST_HOPS, (const char*)&m_iIpTTL, sizeof(m_iIpTTL)))
-               throw CUDTException(MJ_SETUP, MN_NORES, NET_ERROR);
-            //For IPv4mapped-IPv6 accepted connection also set the IPV4 socket.
-            if(0 != ::setsockopt(m_iSocket, IPPROTO_IP, IP_TTL, (const char*)&m_iIpTTL, sizeof(m_iIpTTL)))
-               throw CUDTException(MJ_SETUP, MN_NORES, NET_ERROR);
-         }
-      }   
+          if(m_iIPversion == AF_INET)
```
Comment (maxlovic) [inline srtcore/channel.cpp]: Please add a space after if: `if (m_iIPversion == AF_INET)`

File: srtcore/channel.cpp [hunk 2]
```diff
@@ -201,40 +202,70 @@ void CChannel::setUDPSockOpt()
...
+          {
+              if(0 != ::setsockopt(m_iSocket, IPPROTO_IP, IP_TTL, (const char*)&m_iIpTTL, sizeof(m_iIpTTL)))
```
Comment (maxlovic) [inline srtcore/channel.cpp]: Space after `if`

File: srtcore/channel.cpp [hunk 3]
```diff
@@ -201,40 +202,70 @@ void CChannel::setUDPSockOpt()
...
       if (-1 != m_iIpToS)
       {
-         if(m_iIPversion == AF_INET)
-         {
-            if(0 != ::setsockopt(m_iSocket, IPPROTO_IP, IP_TOS, (const char*)&m_iIpToS, sizeof(m_iIpToS)))
-               throw CUDTException(MJ_SETUP, MN_NORES, NET_ERROR);
-         }
-         else //Assuming AF_INET6
-         {
+          if(m_iIPversion == AF_INET)
```
Comment (maxlovic) [inline srtcore/channel.cpp]: Space after `if`

File: srtcore/channel.cpp [hunk 4]
```diff
@@ -201,40 +202,70 @@ void CChannel::setUDPSockOpt()
...
+          {
+              if(0 != ::setsockopt(m_iSocket, IPPROTO_IP, IP_TOS, (const char*)&m_iIpToS, sizeof(m_iIpToS)))
```
Comment (maxlovic) [inline srtcore/channel.cpp]: Space after `if`

**Comment by jeandube** [comment]
It seems there is a problem with the original fix. A test program running on Linux  3.2.0-29 x86_64 can set TTL but not ToS on an IPV4 socket obtained from an IPv6 listen with IPv4-mapped address enabled.The same test on Linux 4.14 aarch64 can set both TTL and ToS.
The original fix was not tested on a system with IPv6-only enabled  and this is what this PR tries to fix. but it appears the original fix was no portable. Maybe a better way to apply configured TTL and ToS is to set them on the accepted socket and not the ipv6 listening socket 'ipv4-bridge'

**Comment by jeandube** [comment]
Tested on Linux 2.6.37 armv7l and both TTL and ToS are set.

**Comment by ethouris** [comment]
@jeandube We are talking about applying system-specific options set on SRT socket to be spread into underlying UDP sockets. The listener socket shares its underlying UDP socket with all accepted sockets. An accepted socket doesn't get any new UDP socket, so options set through a listener socket will be always effective on all accepted sockets by definition. And, yes, it means that you can't have different options on a listener socket and sockets accepted off this listener.

**Pull request merged**